src/main/java/org/prebid/server/bidder/loyal/LoyalBidder.java

```diff
@@ -0,0 +1,135 @@
+package org.prebid.server.bidder.loyal;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Imp;
+import com.iab.openrtb.response.Bid;
+import com.iab.openrtb.response.BidResponse;
+import com.iab.openrtb.response.SeatBid;
+import io.vertx.core.http.HttpMethod;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderCall;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.loyal.ExtImpLoyal;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.BidderUtil;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+public class LoyalBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpLoyal>> LOYAL_EXT_TYPE_REFERENCE =
+            new TypeReference<>() {
+            };
+
+    private static final String PLACEMENT_ID_MACRO = "{{PlacementId}}";
+    private static final String ENDPOINT_ID_MACRO = "{{EndpointId}}";
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public LoyalBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> requests = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpLoyal ext = parseImpExt(imp);
+                final HttpRequest<BidRequest> httpRequest = createHttpRequest(ext, request);
+                requests.add(httpRequest);
```

Comment:
comparing to the Go version the following logic is missing

```
impExt := reqBodyExt{LoyalBidderExt: reqBodyExtBidder{}}

		if loyalExt.PlacementID != "" {
			impExt.LoyalBidderExt.PlacementID = loyalExt.PlacementID
			impExt.LoyalBidderExt.Type = "publisher"
		} else if loyalExt.EndpointID != "" {
			impExt.LoyalBidderExt.EndpointID = loyalExt.EndpointID
			impExt.LoyalBidderExt.Type = "network"
		}

		finalyImpExt, err := json.Marshal(impExt)
		if err != nil {
			errs = append(errs, err)
			continue
		}

		reqCopy.Imp[0].Ext = finalyImpExt
```

Reply:
Im not sure if I correctly understand It, please check one more time

Reply:
looks like you've done correctly, you can additionally check Go version's tests, it usually helps